subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/CacheBackedTaskHistoryRepository.java

```diff
@@ -379,12 +383,36 @@ private HistoricalTaskExecution loadPreviousExecution(TaskInternal task) {
         return ImmutableSortedSet.copyOf(outputPropertyNames);
     }
 
-    private static ImmutableSet<String> getDeclaredOutputFilePaths(TaskProperties taskProperties, StringInterner stringInterner) {
-        ImmutableSet.Builder<String> declaredOutputFilePaths = ImmutableSortedSet.naturalOrder();
-        for (File file : taskProperties.getOutputFiles()) {
-            declaredOutputFilePaths.add(stringInterner.intern(file.getAbsolutePath()));
-        }
+    private static ImmutableSet<String> getDeclaredOutputFilePaths(final TaskProperties taskProperties, final StringInterner stringInterner) {
+        final ImmutableSet.Builder<String> declaredOutputFilePaths = ImmutableSortedSet.naturalOrder();
+        FileCollectionInternal outputFiles = (FileCollectionInternal) taskProperties.getOutputFiles();
+        outputFiles.visitRootElements(new FileCollectionVisitor() {
+            @Override
+            public void visitCollection(FileCollectionInternal fileCollection) {
+                addAllPaths(fileCollection, declaredOutputFilePaths, stringInterner);
+            }
+
+            @Override
+            public void visitTree(FileTreeInternal fileTree) {
+                addAllPaths(fileTree, declaredOutputFilePaths, stringInterner);
```

Comment:
What will this add for general file trees? The docs say that `FileTree.getFiles()`...

    Returns the contents of this tree as a flattened Set.

Reply:
Yeah, that's exactly what it would do. I don't really know how you would add a `FileTree` as an output that is not a directory tree. I can see people doing something like
```
outputs.files(zipTree('some.jar'))
```
but I wouldn't recommend it.

Do you have any other kinds of file trees in mind?